sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java

```diff
@@ -0,0 +1,216 @@
+        this.receiverBuilder = Objects.requireNonNull(receiverBuilder, "'receiverBuilder' cannot be null");
+        this.processMessage = Objects.requireNonNull(processMessage, "'processMessage' cannot be null");
+        this.processError = Objects.requireNonNull(processError, "'processError' cannot be null");
+        this.processorOptions = Objects.requireNonNull(processorOptions, "'processorOptions' cannot be null");
+        this.asyncClient.set(receiverBuilder.buildAsyncClient());
+        this.sessionReceiverBuilder = null;
+    }
+
+    /**
+     * Starts the processor in the background. When this method is called, the processor will initiate a message
+     * receiver that will invoke the message handler when new messages are available. This method is idempotent i.e
+     * calling {@link #start()} again after the processor is already running is a no-op. Calling {@link #start()}
+     * after calling {@link #stop()} will resume processing messages using the same underlying links or active
+     * sessions. Calling {@link #start()} after calling {@link #close()} will start the processor with new links and
+     * a new set of sessions will be processed.
+     */
+    public synchronized void start() {
+        if (isRunning.getAndSet(true)) {
+            logger.info("Processor is already running");
+            return;
+        }
+        receiveMessages();
+
+        // Start an executor to periodically check if the client's connection is active
+        this.scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
+        scheduledExecutor.scheduleWithFixedDelay(() -> {
+            if (this.asyncClient.get().isConnectionClosed()) {
+                restartMessageReceiver();
+            }
+        }, SCHEDULER_INTERVAL_IN_SECONDS, SCHEDULER_INTERVAL_IN_SECONDS, TimeUnit.SECONDS);
+
+    }
+
+    /**
+     * Stops the message processing for this processor. The receiving links and sessions are kept active and this
+     * processor can resume processing messages by calling {@link #start()} again.
+     */
+    public synchronized void stop() {
+        isRunning.set(false);
+    }
+
+    /**
+     * Stops message processing and closes the processor. The receiving links and sessions are closed and calling
+     * {@link #start()} will create a new processing cycle with new links and new sessions.
+     */
+    public synchronized void close() {
+        isRunning.set(false);
+        if (receiverSubscription.get() != null) {
+            receiverSubscription.get().cancel();
+        }
+        asyncClient.get().close();
+        scheduledExecutor.shutdown();
+    }
+
+    /**
+     * Returns {@code true} if the processor is running. If the processor is stopped or closed, this method returns
+     * {@code false}.
+     *
+     * @return {@code true} if the processor is running.
+     */
+    public synchronized boolean isRunning() {
```

Comment:
Nitpick. Do we need this to be synchronized? This is a read-only operation and `isRunning` is an `AtomicBoolean`.

Reply:
Even though `isRunning` boolean is atomic, the whole method has to be synchronized to provide an accurate state of the processor. Because a thread might have called start() which flips the `isRunning` boolean to true but the call to `receiveMessages()` may not yet be made when another thread tries to call `isRunning()`.